contracts/reference/TokenGatedReferenceModule.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.10;
+
+import {IReferenceModule} from '@aave/lens-protocol/contracts/interfaces/IReferenceModule.sol';
+import {ModuleBase} from '@aave/lens-protocol/contracts/core/modules/ModuleBase.sol';
+import {IERC721} from '@openzeppelin/contracts/token/ERC721/IERC721.sol';
+import {Errors} from '@aave/lens-protocol/contracts/libraries/Errors.sol';
+
+interface IToken {
+    /**
+     * @dev Returns the amount of tokens owned by `account`.
+     */
+    function balanceOf(address account) external view returns (uint256);
+}
+
+/**
+ * @notice A struct containing the necessary data to execute TokenGated references.
+ *
+ * @param tokenAddress The address of ERC20/ERC721 token used for gating the reference
+ * @param minThreshold The minimum balance threshold of the gated token required to execute a reference
+ */
+struct GateParams {
+    address tokenAddress;
+    uint256 minThreshold;
+}
+
+/**
+ * @title TokenGatedReferenceModule
+ * @author Lens Protocol
+ *
+ * @notice A reference module that validates that the user who tries to reference has a required minimum balance of ERC20/ERC721 token.
+ */
+contract TokenGatedReferenceModule is ModuleBase, IReferenceModule {
+    event TokenGatedReferencePublicationCreated(
+        uint256 indexed profileId,
+        uint256 indexed pubId,
+        address tokenAddress,
+        uint256 minThreshold
+    );
+
+    error NotEnoughBalance();
+
+    mapping(uint256 => mapping(uint256 => GateParams)) internal _gateParamsByPublicationByProfile;
+
+    constructor(address hub) ModuleBase(hub) {}
+
+    /**
+     * @dev The gating token address and minimum balance threshold is passed during initialization in data field (see `GateParams` struct)
+     *
+     * @inheritdoc IReferenceModule
+     */
+    function initializeReferenceModule(
+        uint256 profileId,
+        uint256 pubId,
+        bytes calldata data
+    ) external override onlyHub returns (bytes memory) {
+        GateParams memory gateParams = abi.decode(data, (GateParams));
+
+        // Checking if the tokenAddress resembles ERC20/ERC721 token (by calling balanceOf() function)
+        (bool success, bytes memory result) = gateParams.tokenAddress.staticcall(
+            abi.encodeWithSignature('balanceOf(address)', address(this))
+        );
+        // We don't check if the contract exists cause we expect the return data anyway
+        if (gateParams.minThreshold == 0 || !success || result.length == 0 || result.length > 32)
+            revert Errors.InitParamsInvalid();
```

Comment:
This must be tested to make sure I'm not wrong, but I guess we can do:

```solidity
uint256 internal constant UINT256_BYTES = 32;
```
And then:
```suggestion
        if (gateParams.minThreshold == 0 || !success || result.length != UINT256_BYTES)
            revert Errors.InitParamsInvalid();
```

But I'm assuming the answer is always padded to fit the return type (uint256),

Reply:
Yes, that's correct - even `returns (uint8)` length is 32 bytes.